### DISCUSSION TITLE
캐싱이 무엇이고 사용하면 좋은 점이 무엇일까? 또 안 좋을 때는 언제인가?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#35] [OOP]
.

**Comment by le2sky** [comment]
으음 분류가 애매하네요 이건... OOP는 아닌 것 같구.. 
네오 학습 테스트에서는 클린 코드 파트에 있던 것 같네요.
아니면 캐싱은 성능에 대한 이야기니, 자바에 있으면 좋을까요?

**Reply by alstn113** [reply]
어디 넣을지 애매하더라구요
그냥 객체 캐싱이라서 OOP라고 생각하고 넣었습니다.

**Comment by alstn113** [comment]
캐싱(Caching)은 데이터나 계산 결과를 임시로 저장하여 나중에 동일한 요청이 발생할 때 다시 계산하지 않고 저장된 값을 사용하는 것을 말합니다. 캐싱은 주로 시간이 많이 걸리거나 자주 요청되는 작업의 성능을 향상시키기 위해 사용됩니다.

캐싱의 장점은 다음과 같습니다.
1. 이전에 계산한 결과를 재사용함으로써 시스템의 응답 시간을 줄일 수 있습니다. 데이터나 연산 결과를 메모리나 디스크에 저장하고 재활용함으로써 요청에 대한 응답 시간을 크게 줄일 수 있습니다.
2. 반복적인 작업을 다시 수행하지 않아도 되므로 시스템 자원을 절약할 수 있습니다. 특히 복잡한 계산이나 데이터베이스 쿼리와 같은 고비용 작업의 반복 수행을 피할 수 있습니다.

캐싱을 사용하는 것이 안 좋은 경우도 있습니다.
1. 캐시된 데이터가 업데이트되지 않을 경우 데이터 일관성 문제가 발생할 수 있습니다.
2. 캐시된 데이터가 많아지면 메모리 부족 문제가 발생할 수 있습니다.
3. 캐시된 데이터가 자주 변경되거나 요청이 다양하거나 빈번하게 발생하는 경우에는 캐싱을 적용하기 어려울 수 있습니다.

캐싱은 성능 향상과 자원 절약을 위해 매우 유용하지만, 적절한 시나리오에서만 사용하는 것이 중요합니다.
데이터의 업데이트 빈도와 데이터의 중요도를 고려하여 적절한 캐싱 전략을 수립해야 합니다.

**Comment by GIVEN53** [comment]
캐싱은 반복적으로 사용되는 데이터나 연산 결과를 임시로 저장하는 메커니즘이다. 빠른 응답 시간과 시스템 성능 향상을 위해 사용된다.

### 장점
1. 응답 시간 감소
    - 반복적으로 사용되는 데이터를 메모리나 다른 빠른 저장소에 저장하므로, 데이터에 접근하는 시간을 단축시켜 응답 시간을 줄일 수 있다.
2. 시스템 성능 향상
    - 빈번하게 사용되는 데이터나 연산 결과를 메모리에 저장해서 서버의 부하를 줄이고 시스템 전체적인 성능을 향상시킬 수 있다.

### 단점
1. 메모리 사용량 증가
    - 캐시된 데이터가 많아지면 메모리 사용량이 증가할 수 있다.
2. 캐시 일관성 관리
    - 캐시된 데이터와 원본 데이터의 일관성을 유지하기 위해 적절한 캐시 제어 및 관리가 필요하다.
3. 데이터 무효화 오버헤드
    - 데이터의 변경이 발생하면 관련된 캐시를 무효화해야 한다.
4. 캐시 미스
    - 캐시에서 데이터를 찾지 못하고 원본 데이터를 다시 가져와야 하는 경우 캐시 미스가 발생한다. 이 때 응답 시간을 느리게 만들 수 있다.
